Run AnalyzeChangeJob’s diffs in parallel

bensheldon opened the pull request:
Uses Concurrent Ruby's Promises: https://github.com/ruby-concurrency/concurrent-ruby/blob/master/docs-source/promises.in.md

Closes #420

Mr0grog commented on app/jobs/analyze_change_job.rb:
```diff
@@ -73,39 +73,44 @@ def perform(to_version, from_version = nil, compare_earliest = true)
   def analyze_change(change)
+    text_diff_promise = Concurrent::Promise.execute { Differ.for_type!('html_text_dmp').diff(change)['diff'] }
+    source_diff_promise = Concurrent::Promise.execute { Differ.for_type!('html_source_dmp').diff(change)['diff'] }
+    links_diff_promise = Concurrent::Promise.execute { Differ.for_type!('links_json').diff(change)['diff'] }
```
The docs say `Concurrent::Promise` has been deprecated and replaced with `Concurrent::Promises` (singular vs. plural for maximum hard-to-spot-ness). The new module has `#future` instead of `#execute`, but it looks like you’d otherwise use it the same here.

Since we create these and resolve them in the same scope and before touching `change` again, it’s probably not important, but maybe more safety-conscious to pass `change` as an argument.

Also maybe worth making a method for this like you did `diff_changes`?

```rb
schedule_diff(type, change)
  Concurrent::Promises.future(type, change) { |t, c| Differ.for_type!(t).diff(change)['diff'] }
end

bensheldon replied:
Thanks for digging into the `Concurrent::Promise` docs and pointing me in the right direction. 👍

bensheldon commented:
@Mr0grog I added the simplest test case I could think of for `AnalyzeChangeJob`. Could you take a look at let me know what I've gotten wrong about it. My main goal intent was to test through the `Differ`, though I'm not quite sure if I got the setup correct. For example:

- I use the `changes` fixture. I wasn't sure whether I should use that fixture, or just make up a new pair of uuids because it appears that a new `Change` resource is created within this job.
- I hope my fake diff data is of the correct shape. I couldn't find an example of the expected data so I just molded something that fit the code.

Mr0grog commented on test/jobs/analyze_change_job_test.rb:
```diff
@@ -0,0 +1,31 @@
+          [-1, 'subtraction']
+        ]
+      }
+    end
```
Yep, this is a good mock. There’s more stuff in the envelope object that wraps the diff, but it’s generally not relevant to this job. But if you’re curious, a more complete example would be:

```js
{
  "version": "1.2.3+hash",
  "type": "links",
  "change_count": 2,
  // other fields specific to selected diff type...

  // Diff is usually this format, but some diffs return something else, like
  // an HTML string (e.g. html_token). All the ones relevant for this test
  // follow this format, though. :)
  "diff": [
    [1, 'addition'],
    [-1, 'subtraction']
  ]
}
```

bensheldon replied:
Thanks! I updated the hash to match a little better in case I have this question again in the future.

Mr0grog commented on test/jobs/analyze_change_job_test.rb:
```diff
@@ -0,0 +1,31 @@
+
+class AnalyzeChangeJobTest < ActiveJob::TestCase
+  test 'produces change with annotations' do
+    change = changes(:page1_change_1_2)
```
> I use the `changes` fixture. I wasn't sure whether I should use that fixture, or just make up a new pair of uuids because it appears that a new `Change` resource is created within this job.

Ah, so this job is largely responsible for creating changes, so it might be better to create a new pair of UUIDs to analyze. (Technically the job makes an annotation, which makes a relevant change if one doesn’t already exist. But most of the time this job runs, that change doesn’t exist yet. A change mainly exists as a way to tie together and summarize multiple annotations.)

bensheldon replied:
Done!

bensheldon commented:
@Mr0grog I addressed your requested changes:

- I added a test, and improved them to create a new set of versions to mock against
- I fixed my usage of `Concurrent::Promises` and added it to the Gemfile.
